### PR TITLE
Unbreak input profile : add log profile

### PR DESCRIPTION
**Pull request by aurelienpierre**
_Warning : This is my first code contribution in dt and the first time I use C and work with GUIs._

This adds a new mode of tone-mapping before color profile correction, using the same approach as video encoding Log profiles (S-log, N-log, etc.). Details and example can be seen here : https://discuss.pixls.us/t/solving-dynamic-range-problems-in-a-linear-way/9006

It retains the previous gamma method and takes care of the legacy params.

It adds an auto solver to automatically compute the pest settings, namely the dynamic range, according to the target grey and the measurements from the color sampler.

It adds masking/blending options and the ability to add in styles.

The OpenCL code is given but does not work at this point. I need help to debug it. OpenCL is enabled for the gamma method and disabled for now on the log method. See https://www.mail-archive.com/darktable-dev@lists.darktable.org/msg03355.html

### CODE REVIEW

File: src/iop/profile_gamma.c [hunk 1]
```diff
@@ -304,33 +780,149 @@ void cleanup_global(dt_iop_module_so_t *module)
...
+  dt_bauhaus_widget_set_label(g->gamma, NULL, _("gamma"));
+  gtk_box_pack_start(GTK_BOX(vbox_gamma), g->gamma, TRUE, TRUE, 0);
+  gtk_widget_set_tooltip_text(g->gamma, _("gamma exponential factor"));
   g_signal_connect(G_OBJECT(g->gamma), "value-changed", G_CALLBACK(gamma_callback), self);
+
+  gtk_widget_show_all(vbox_gamma);
+  gtk_stack_add_named(GTK_STACK(g->mode_stack), vbox_gamma, "gamma");
+
+
+  /**** LOG MODE ****/
```
[the remark this answers is not in the document]
Comment (aurelienpierre) [reply src/iop/profile_gamma.c]: What stuff ? What is mislabelled ? I don't understand what you mean

[the remark this answers is not in the document]
Comment (LebedevRI) [reply src/iop/profile_gamma.c]: Most of the defaults for these sliders in`"camera & signal properties"`.
You should be able to get most of that from the actual black/white levels.
Else i suspect the name is unrepresentable.

[the remark this answers is not in the document]
Comment (aurelienpierre) [reply src/iop/profile_gamma.c]: It's not black level as in the exposure module, it's the limit between black (signal) and noise. Using a log function, values very close to zero (thus noise) get amplified much more than correct values. We need a threshold to cut out very small values before sending them in the log. This depends on the sensor, ISO value, previous denoising and demosaicing, and exposure compensation (gain).

The linear factor helps to center the dynamic range such that black are at - DR / 2, white are at +DR / 2, and the target middle grey is at 0 EV, thus 50 %. It should be measured or estimated by optimization.

[the remark this answers is not in the document]
Comment (sobotka) [reply src/iop/profile_gamma.c]: Again, I'd leave things as a base log. You shouldn't be having any noise amplification at all, and if you are, the code is busted.

Base logs have been around longer than DSLRs. They work.

[the remark this answers is not in the document]
Comment (aurelienpierre) [reply src/iop/profile_gamma.c]: > Again, I'd leave things as a base log. You shouldn't be having any noise amplification at all, and if you are, the code is busted.
> 
> Base logs have been around longer than DSLRs. They work.

I'm sorry, but in dt, this module can take negative values which are going to output NaN or zero division errors. I trust DSLR manufacturer can normalize the values before the log, taking account of the noise level, but we have to take whatever crap the previous modules in the pixelpipe give us.

> Most of the defaults for these sliders in`"camera & signal properties"`.
> You should be able to get most of that from the actual black/white levels.
> Else i suspect the name is unrepresentable.

You were actually right, looking again at the maths, these parameters are dependent of the dynamic range so I just removed them throm the UI.

[the remark this answers is not in the document]
Comment (LebedevRI) [reply src/iop/profile_gamma.c]: > You were actually right, looking again at the maths, these parameters are dependent of the dynamic range so I just removed them throm the UI.

... are you still confident that dynamic range won't be auto-calculatable the way i suggested in the first comment in this thread?

[the remark this answers is not in the document]
Comment (aurelienpierre) [reply src/iop/profile_gamma.c]: no, this dynamic range is the output dynamic range. You can see that as a target contrast. It can be optimized to fit the whole [0; 1]  range at the output, considering the input extrema (that's the purpose of the last slider + autofix_callback function), but there is no way you could guess it from hadrware properties (and, again, it depends on what you did n the previous modules).

**Comment by aurelienpierre** [comment]
Please find a full step-by-step tutorial of this module with examples here : https://discuss.pixls.us/t/solving-dynamic-range-problems-in-a-linear-way/9006/63?u=aurelienpierre

**Comment by aurelienpierre** [comment]
OpenCL is now working

**Comment by aurelienpierre** [comment]
> The existing `unbreak input profile` module is 'designed' just for that - to cope with
> some existing **broken** ICC color profiles.

An ICC profile is calibrated on a color chart. These semi-reflective charts have patches with L values ranging from 18 (black) to 96 (white). What happens between L = 0 and L = 18 are extrapolations. In a modern RAW file, from a 10-14 EV of dynamic range sensor, exposed to not clip the highlights, you get 75 to 90 % of your pixels below L = 18. When you feed that to the input ICC correction, you are practically hoping for the best, because you are outside its validity boundaries. So, yes, you are unbreaking it. The beauty of the log method is you can select the average luminance of the non-corrected RAW and have it remapped to the center of the histogram, hence the center of the validity domain of the ICC profile. It consistently gives symmetrical histograms centered on 50 %.

Besides, here is the current gamma transfer function for gamma = 2.2 (linearity = 0) :
image

and here is the log version transfer function (dynamic range = 6.7 EV, grey = 1)
image

It's not like day and night, nor it is an artistic transformation. It's similar to what you see in the basecurve module, except this is parametric and doesn't give highlights over-saturation.

> Also, i find it kind-of misleading to compare this with `video encoding Log profiles (S-log, N-log, etc.)`.

Well, it's the same function for the exact same purpose : compressing the dynamic range. Ok, there is no destructive encoding in the next step, but that's anecdotal.

> Also, rangecheck: the code needs to cope with _any_ input.  I saw there is some handling for negative values, but i have no clue if that is sufficient.

The handling of out-of-boundaries values is as follow: 

- for any input > 2^(-dynamic range), typically that is in [0.00006 ; 0.002], process (input + 0.000061035) and subtract 0.000061035 on the output.

- for any input =< 2^(-dynamic range), process 2^(-dynamic range).

**Comment by aurelienpierre** [comment]
The gamma version of the unbreak profile uses a lookup table contructed somehow with integers for inputs < 1, I have no idea what's going on in there and there are no comments. 

I don't even understand how you end up with negative numbers at that point… That makes no sense.

**Comment by LebedevRI** [comment]
And as is tradition with all these color-related threads, people are talking past each other.

**Comment by LebedevRI** [comment]
> > What comment would you want to see there ^ ?
> 
> That's what is run. But the coef that are put in the `d->table` look-up table are constructed like this :

Correct, some look-up-table for the normal values in `[0.0, 1.0]` range, with resolution of `0x10000` (`65536`) samples.

> So I don't get **where** the gamma function does clamp its input

From my previous comment
https://github.com/darktable-org/darktable/blob/d99df3190faffffb0c66f813d5d5461ff6b89d37/src/iop/profile_gamma.c#L157-L161

"if the pixel value is less than `1.0`, then multiply it by the resolution of the look-up-table (`0x10000` (`65536`)), lossy convert to integer, clamp to [0, `size of LUT-1`], look up in LUT in that column, and set the pixel to that value. else, if the pixel is > 1.0, evaluate some equation with x = pixel value to get the approximate curve value"
Now, i have to agree, i'm not sure why, but i was rather sure that /some/ modules were doing this special handling not just for the `>1.0` pixels, but for `<0.0` too, but i'm not seeing it.

**Comment by aurelienpierre** [comment]
> Now, i have to agree, i'm not sure why, but i was rather sure that /some/ modules were doing this special handling not just for the `>1.0` pixels, but for `<0.0` too, but i'm not seeing it.

So we are on the same page here. A lossy conversion from a negative float to a 16-bit integer seems to be 0…

**Comment by edgardoh** [comment]
Hi,

I've been playing a bit and here are my first impressions:

1. Is not that easy to get good results, with well exposed images I usually get artefacts, so I have to go to the exposure, make some changes, go back here, adjust it, etc.
2. Once I get a flat image with no artifacts, sometimes when using the tone curve or the contr-brigh-sat I get artifacts again, so I have to go back to the beginning.
3. I managed to get rid of the artifacts, now I can play, but is not that easy to restore the contrast/saturation. OK, that is maybe because of my editing skills, but at least in part I think is because all the things I have done to get rid of the artifacts.

Now, having done all that, the end result is nicer than I could get with the base curve, and it give me a better base to continue my edits. So, is there anything that can be done about the artifacts? Have in mind that I see this as an end user, I don't know the first thing about the subject.

**Comment by aurelienpierre** [comment]
Hi Edgardo,

Thanks for your test. I'm a bit surprised, since I get that module to work in just a few steps. What sort fo artefacts do you get ? Have you followed the tutorial here: https://discuss.pixls.us/t/solving-dynamic-range-problems-in-a-linear-way/9006/63?u=aurelienpierre

>     3\. I managed to get rid of the artifacts, now I can play, but is not that easy to restore the contrast/saturation. OK, that is maybe because of my editing skills, but at least in part I think is because all the things I have done to get rid of the artifacts.

Keep in mind that my other PR #1734 is intended to be exactly the next step to this module. With the fulcrum contrast and the power (mid-tones) adjustement, I do 80 % of the editing in 30 s with just 2 modules.

**Comment by edgardoh** [comment]
I did follow it. BTW, great tutorial, I've couldn't done it without it.

I'll see if I can find an image that I can share with those issues, so you
have a better idea, but by artifacts I mean red/blue/green pixels usually
on dark areas. Even if they are not present in this module the show by just
enabling the levels or the contr-brigh-sat, so maybe (and this is a wild
guess) something is out of gamut.

I get that this is my first interaction with the module, with practice
things will go faster and results better. And since I already manage to get
good results there is no doubt that this module has potential.
What I would like is to minimize the chance of those artifacts and an even
easier way to get nice results.

Maybe this is how the algorithm works and there is nothing you can do about
it, but if you can identify the problem maybe a message or warning "there
are pixels out of gamut (or whatever the problem is)" so we don't need to
check the entire image each time.

The other thing is to go to the exposure and back here, so maybe is
possible to make this a new module, place it (on the pipe) before the
exposure, and add the black point/exposure here. This way you have control
over it, and an auto button can be created that sets the black
point/exposure, uses the entire image for the middle grey and dynamic
range, and the darker area for the black exposure. This should give nice
results.


2018-09-29 14:07 GMT-03:00 Aurélien PIERRE <notifications@github.com>:

**Comment by edgardoh** [comment]
I did some more testing and it is a gamut issue.

On an image with no clipped nor out-of-gamut areas I select the entire image for middle grey and dynamic range and the darkest area for black exposure.
I get some out-of-gamut areas.
I can't say if this is a problem or not, you know more than me on this subject, but if I use the levels, contr-brigh-sat or tone curve in lab I get artifacts. I have used those modules with out-of-gamut input with no artifacts, so something is different here.
I also know that those modules make assumptions on the input data and they shouldn't, so maybe a rgb mode should be added to them (the tone curve already has it)

**Comment by aurelienpierre** [comment]
If these out-of-gamut issues happen in low-lights, you need to decrease the black level in the exposure module. That's a drawback of the log, very low luminance values get lowered even more, so you need to add an offset at the input. Which is the black level.

**Comment by edgardoh** [comment]
OK, but a warning will be nice, maybe a label "you have x pixels out-of-gamut, decrease the black level". 
Also the auto button will be nice, I'll probably use the same workflow for most images, so if it can be automated it will be great. And it will free the exposure for other uses, right now if I create a new instance it will mess with the edits.

**Comment by aurelienpierre** [comment]
> maybe a label "you have x pixels out-of-gamut, decrease the black level

That would suppose to count the underexposed pixels during the processing loop. I see how to do that in pure C but, in OpenCL, it would need to parse the whole image before or after the processing, so it would defeat the purpose of vectorizing for performance. I have to think about that…

> Also the auto button will be nice

I tried to do that at first, the problem is, as the black level auto-detection is super sensitive to noise (no algo can guess if a low value is noise or a true black), I often ended up tweaking the "optimized" auto-computed parameters anyway, so I just removed the full auto mode and went for a trade-off where you give the algo hints of what is black and what is grey. I feel it's better to give that responsibility to the user rather than creating frustration with a partially-working auto mode.

**Comment by edgardoh** [comment]
-The label should be updated only when a setting has changed, and run only
on the preview pipe, so performance won't be an issue. You can have a CPU
function that go though all the image and count the pixels, since there's
user interaction it won't probably be noticed.

-For the auto button, maybe decreasing the black level until there's no
out-of-gamut pixels? That's my goal anyway. Or having a slider with a
percent of out-of-gamut pixels allowed. This is again user interaction, so
performance is not an issue.

-The position in the pipe is more important to me anyway, this forces me to
have only one instance of the exposure and to use it only to set the black
level. That's a big restriction.


2018-09-29 21:52 GMT-03:00 Aurélien PIERRE <notifications@github.com>:

**Comment by aurelienpierre** [comment]
I have added a full auto button, for the better and the worse.

I still have to find out in which structure is hidden the roi_in piece so that I can parse it in `commit_params` function.

**Comment by edgardoh** [comment]
The idea of the auto button was based on this module to be moved before the
exposure, and the black and exposure setting added to this very module, so
the can also be optimized. That's obviously not going to happen, and with
this current version, as you already said, results can be not good, so
never mind.

2018-09-30 3:10 GMT-03:00 Aurélien PIERRE <notifications@github.com>:

**Comment by aurelienpierre** [comment]
I think I have fixed the auto-optimizer. Provided that there is not too much noise, it should consistently remap the Lab luminance between 18 and 96. Could you test it ?

**Comment by rawfiner** [comment]
Hi
I am (for now) only a user of darktable, so I guess my opinion is less important than the ones of the main developers, but here it is.
I have tested the latest version of this. I personally think that this pull request adds a great tool.
Combined with the color balance modification, it makes the handling of complex images with backlight much easier.
I previously struggled to get some correct results on images with wide dynamic. With these modifications, I managed to get nice results a lot faster, and the results were better.
I think this particular pull request has improved a lot, and it is now very easy (and very efficient) to use the module, with all the color pickers and the autotune button.
I tested it on raw images that were noisy as well, and the module succeeded to find good values despite of the noise.
Long story short, in my humble opinion, as a user, this module is a very valuable addition.
I also would like to suggest that people who tested it a while ago give it another chance, to take into account the changes that Aurélien made recently.

**Comment by jpg54** [comment]
As a user, I can't discuss programming considerations but I find that Aurélien's algorithms present another way to process his RAW. Moreover, it does not prevent the use of old algorithms and leaves the user the choice a bit like in local contrast.

**Comment by TurboGit** [comment]
Aurélien, when you have some time can you fix the merge conflicts? I'll have a look, all those nice comments push me to see more about it. Thanks.

**Comment by aurelienpierre** [comment]
@TurboGit done !

**Comment by TurboGit** [comment]
Still some merge conflicts? Have't you forgotten to push some modifications?

**Comment by aurelienpierre** [comment]
It was `src/common/colorspaces_inline_conversions.h` that was in conflict, right ? It should be fixed, I have rebased.

**Comment by TurboGit** [comment]
Yes, but you need to push in aurelienpierre:color-grading. The conflict is still there.

**Comment by TurboGit** [comment]
Aurélien, I will close this PR. There is 60 commits, many merges with master, it is possible to merge locally but impossible to rebase due to conflicts.

I have squashed all those commits into a single one and will propose a new PR with this. A cleaner start for the integration.
